File: workspaces-good/packages/agent_calculator_orchestrator.ts
import { CalculatorAggregate } from "calculator-shared/src/contract_calculator_aggregate";
import { CalculatorProtocol } from "calculator-shared/src/contract_calculator_protocol";
import { ExpressionVO } from "calculator-shared/src/taxonomy_expression_vo";
import { OperationVO } from "calculator-shared/src/taxonomy_operation_vo";
import { ResultVO } from "calculator-shared/src/taxonomy_result_vo";

// ─── Block 1: Struct Definition ───────────────────────────

export interface CalculatorOrchestratorDeps {
  addition: CalculatorProtocol;
  subtraction: CalculatorProtocol;
  multiplication: CalculatorProtocol;
  division: CalculatorProtocol;
}

// ─── Block 2: Protocol Implementation ─────────────────────

export class CalculatorOrchestrator implements CalculatorAggregate {
  private deps: CalculatorOrchestratorDeps;
  private _history: ResultVO[] = [];

  constructor(deps: CalculatorOrchestratorDeps) {
    this.deps = deps;
  }

  delegate(expr: ExpressionVO): ResultVO | null {
    const analyzerMap: Partial<Record<OperationVO, CalculatorProtocol>> = {
      [OperationVO.Add]: this.deps.addition,
      [OperationVO.Subtract]: this.deps.subtraction,
      [OperationVO.Multiply]: this.deps.multiplication,
      [OperationVO.Divide]: this.deps.division,
    };
    const analyzer = analyzerMap[expr.op];
    if (!analyzer) return null;
    const result = analyzer.evaluate(expr);
    if (result) this._history.push(result);
    return result;
  }

  history(): ResultVO[] {
    return [...this._history];
  }
}

File: workspaces-good/modules/agent_calculator_orchestrator.py
from typing import List, Optional

from shared.src.contract_calculator_aggregate import CalculatorAggregate
from shared.src.contract_calculator_protocol import CalculatorProtocol
from shared.src.taxonomy_expression_vo import ExpressionVO
from shared.src.taxonomy_operation_vo import OperationVO
from shared.src.taxonomy_result_vo import ResultVO


# ─── Block 1: Struct Definition ───────────────────────────

class CalculatorOrchestratorDeps:
    def __init__(
        self,
        addition: CalculatorProtocol,
        subtraction: CalculatorProtocol,
        multiplication: CalculatorProtocol,
        division: CalculatorProtocol,
    ):
        self.addition = addition
        self.subtraction = subtraction
        self.multiplication = multiplication
        self.division = division


# ─── Block 2: Protocol Implementation ─────────────────────

class CalculatorOrchestrator(CalculatorAggregate):
    def __init__(self, deps: CalculatorOrchestratorDeps):
        self._deps = deps
        self._history: List[ResultVO] = []

    def delegate(self, expr: ExpressionVO) -> Optional[ResultVO]:
        analyzer_map = {
            OperationVO.ADD: self._deps.addition,
            OperationVO.SUBTRACT: self._deps.subtraction,
            OperationVO.MULTIPLY: self._deps.multiplication,
            OperationVO.DIVIDE: self._deps.division,
        }
        analyzer = analyzer_map.get(expr.op)
        if analyzer is None:
            return None
        result = analyzer.evaluate(expr)
        if result is not None:
            self._history.append(result)
        return result

    def history(self) -> List[ResultVO]:
        return list(self._history)
